cmd/ndm_daemonset/controller/blockdevicestore.go

```diff
@@ -68,67 +71,49 @@ func (c *Controller) CreateBlockDevice(blockDevice apis.BlockDevice) error {
 		return err
 	}
 
-	/*
-	 * Update might failed due to to resource version mismatch which
-	 * can happen if some other entity updating same resource in parallel.
-	 */
-	err = c.UpdateBlockDevice(blockDevice, nil)
-	if err == nil {
-		return err
-	}
 	klog.Error("Update to blockdevice object failed: ", blockDevice.ObjectMeta.Name)
 	return nil
 }
 
+func boolPtr(b bool) *bool {
+	return &b
+}
+
 // UpdateBlockDevice update the BlockDevice resource in etcd
-func (c *Controller) UpdateBlockDevice(blockDevice apis.BlockDevice, oldBlockDevice *apis.BlockDevice) error {
+func (c *Controller) UpdateBlockDevice(blockDevice apis.BlockDevice) error {
 	var err error
 
-	blockDeviceCopy := blockDevice.DeepCopy()
-	if oldBlockDevice == nil {
-		oldBlockDevice = blockDevice.DeepCopy()
-		err = c.Clientset.Get(context.TODO(), client.ObjectKey{
-			Namespace: oldBlockDevice.Namespace,
-			Name:      oldBlockDevice.Name}, oldBlockDevice)
-		if err != nil {
-			klog.Errorf("eventcode=%s msg=%s : %v, err:%v rname=%v",
-				"ndm.blockdevice.update.failure",
-				"Failed to update block device : unable to get blockdevice object",
-				oldBlockDevice.ObjectMeta.Name, err, blockDeviceCopy.ObjectMeta.Name)
-			return err
-		}
-	}
-
-	blockDeviceCopy = mergeBlockDeviceData(*blockDeviceCopy, *oldBlockDevice)
+	klog.Infof("Blockdevice on API server %v", blockDevice)
+	patch := client.MergeFromWithOptions(blockDevice.DeepCopy())
 
-	err = c.Clientset.Update(context.TODO(), blockDeviceCopy)
+	err = c.Clientset.Patch(context.TODO(), &blockDevice, patch)
```

Comment:
If I'm not wrong `UpdateBlockDevice` is used to update bd with the new values provided in the `blockDevice` parameter right? But since we are using a copy of the `blockDevice` as the base in `MergeFromWithOptions` and then sending the same `blockDevice` in `ClientSet.Patch` won't the resultant patch have no changes? I could be wrong about the flow here. Let me know if I'm missing something.

Reply:
I'm not sure about the best way to pick up changes here. If we use `Update` call then that results in a conflict. Even after I fetch it right before making the `Update` call . All the test cases are failing because of this, the only way I found which solves this is in the last commit (Fix ZFS test case) .

Reply:
Interesting. I tried debugging the `TestUpdateDevice`. I found that the final patch data being generated is `[]byte{123,125}`, which corresponds to `{}`. So looks like the device won't be updated since the patch is empty. The test also doesn't seem to be checking for this. I think the tests need to be updated too. Here's a screenshot of the debugging session (focus on the value of `data`):
![Screenshot from 2021-05-20 15-02-40](https://user-images.githubusercontent.com/31250892/118955788-a6c5db80-b97c-11eb-89da-06e0a795adac.png)

Reply:
@RealHarshThakur I think the concern raised by @z0marlin will happen. the patch is not having any changes. That is the reason the call is not erroring out.

Reply:
@RealHarshThakur Can you revisit the logic for the Update method, we need to find a suitable method to do it. I will also test it out locally.

Reply:
>  Also @z0marlin there will be changes in the patch in some cases, because the base object and etcd can differ. (though not always)

I think the above statement from me is wrong. If using the same object in the patch call and patch, there will be no changes. We have to see how a 3 way merge patch works.

Reply:
True @akhilrm. Going through the merge patch code, even I suspect the patch data will always be empty since the patch is by comparing the base object set in the `MergeFromWithOptions` call, and the object sent in the `Patch` (the second parameter). These two are same since the former is a deepcopy of the latter in this code.

Reply:
@akhilerm Worst case scenario: I think we should `Update` or `Patch` the resource in the same function and remove/minimize the use of this function itself.  I did that in last commit (ZFS test case). WDYT?